[WIP] Simplified Deploy

dilatebrave opened the pull request:
Fix: #154

**Description**

Step 1:
* [x] [Dropdown] Select the API which user would like to use. Example: Binance/Bittrex. This would be defaulted to Binance
* [x] [Dropdown] Select the symbol which is returned from the Binance exchangeInfo API

Step 2:
* [x] [Input] Pre-fill the Decimals field with the decimals reported from Binance for the above selected symbol
* [ ] [Input] Price Cap and Price Floor -- Default to 50% above and below the current value for the above selected symbol
* [ ] [Input] Price Qty Multiplier : TBD

Step 3:
* [x] [Input] Standardize the Name as per the standards mentioned here -- MARKETProtocol/MARKETProtocol#44
* [x] [Date] Expiration time -- Default it to 30 days
* [x] [GasPriceField] (It's an existing component)

In the background:
* [ ] Type of Oracle Query (Re-use the existing component. Default value: URL)
* [ ] Generate Oraclize Query based on the above selected symbol as well as the input of the Price fields which will be the value of the Oraclize Query


**Definition of Done**

* [x] Create needed UX flow for Simplified Deploy
* [ ] Plan to interact with core team to ensure expectations are met, plan for a few revisions
* [ ] For integration of Binance exchangeInfo API, please use the existing code from this [PR] (#86). Feel free to reach out to @eswarasai in case any help required here
* [ ] Re-use components as much as possible with few tweaks in order to avoid duplication. We strictly follow DRY policy
* [ ] Add/Update all needed/broken tests for all the code written

pelsasser commented:
re Qty Multiplier -

Example

ADA / ETH - .00034795 on binance right now.

We will be converting this into an integer using Price Decimal Places
.00034795 --> 00034795.0  (8 decimal places)

And for every integer price movement (34795 to 34796 for example) we want to determine how much collateral is needed per traded qty. 

If the price multiplier was 1000000000000000000 then for each change the value would be worth 1 ETH (way too big!)

If the price multiplier was 10000000000 then for each change the value would be worth 0.00000001 ETH or 10 gwei which is the same as it is on binance.

(see here for help with units and ETH - https://etherconverter.online/)

So for all the pairs denominated in ETH we can use the number of decimal places and the number of wei in ETH to create the correct multiplier. 

See why we need a simplified deployer  ;)

rojosnow commented:
@pelsasser So, using ADA / ETH - .00034795, if I'm a maker and want to execute a contract, quantity 1000, with a ceiling of .00034895 ADA, I'd need 1000 (quantity) * 100 (price difference)  * 10 gwei = 1,000,000 gwei or 1 micro of ETH?

pelsasser commented:
@rojosnow - it doesn't matter if maker or taker, but more specifically if buyer or seller.

See https://github.com/MARKETProtocol/MARKETProtocol/blob/b086c6d34396fcb15f9e90cb50341f59ccf72ef5/contracts/libraries/MathLib.sol#L125

So if a buyer we use price difference from traded price to floor, and if seller from traded price to cap.

otherwise your logic is correct.

rojosnow commented:
@pelsasser You can tell I like to think about things in gains! ;) Got it!

eswarasai commented:
@dilatebrave -- This looks really good to me. Please find the changes needed below :

- The order of dropdown for the `Deploy Contract` on the Navbar Menu should be `Quick Deploy`, `Simplified Deploy` and `Guided Deploy`

- We need to hide the helper text for Simplified flow but should be displayed for the Guided flow for all the input components

- Fix the default value of `Price Cap` to be integer. PFA screenshot of error below
<img width="1296" alt="screen shot 2018-05-19 at 1 54 53 am" src="https://user-images.githubusercontent.com/5172086/40256333-b7598b44-5b07-11e8-8675-ce8dc9e24560.png">

- Update the default value of `Price Quantity Multiplier` as per mentioned [above](https://github.com/MARKETProtocol/dApp/pull/156#issuecomment-390234303)

- Reverse the order of `Expiration Time` and `Name` fields in Step 3 and also fix the `undefined` in the standardized `Name` value
<img width="1320" alt="screen shot 2018-05-19 at 1 58 49 am" src="https://user-images.githubusercontent.com/5172086/40256577-8e86224e-5b08-11e8-92a7-884b270c81a0.png">

- Lastly, might want to bump up the coverage a bit more in order for us to approve the PR

Thanks 👍

dilatebrave commented:
@eswarasai -- it's done

eswarasai commented:
@dilatebrave -- Thanks for resolving them so quickly. I just have couple of minor changes :

- Remove the `package-lock.json` diff
- Fix the below console error

<img width="498" alt="screen shot 2018-05-22 at 12 09 36 am" src="https://user-images.githubusercontent.com/5172086/40323991-b80f3c06-5d54-11e8-8e4e-cc93bea47e27.png">

Also please do rebase your branch with that of `MARKETProtocol/dApp:develop` and amend these changes to your last commit.

@pelsasser -- Is there anything else that'd you like to add to this PR?

pelsasser commented:
@dilatebrave - I am amazed at your ability to program this up so quickly and without many questions to us! It isn't exactly straight forward and this look awesome!

A few questions / comments (some of these will probably end up being issues we create for the next pass on this)
1. Do you think we even need to display Price Decimal Places to the user? perhaps this is confusing and should not really be edited. (same with Price Qty Multiplier, probably more confusing then helpful)?
1. In the same vain should we display the Floor and Cap in decimals representation and only behind the scenes modify them to integers.
1. Perhaps displaying to the user the current price would be helpful as well
1. Eventually we should add pairs that are denominated in USD / stable coins as well, we will have to modify some of these calculations, but it appears this shouldn't be a problem.

@eswarasai @dilatebrave nice job on this!

@dilatebrave - I hope you will continue to help us I am guessing you have some background that lends itself to our project and issues like these!  Please join our [discord](https://marketprotocol.io/discord) if you haven't already.

dilatebrave commented:
@eswarasai -- i fixed the bug, used backend api and rebased the pr!

@pelsasser -- i glad to help if i can, my discord is dilatebrave too

> Do you think we even need to display Price Decimal Places to the user? perhaps this is confusing and should not really be edited. (same with Price Qty Multiplier, probably more confusing then helpful)?

Yes, i think it is confusing too. I have to read the helper text over and over to understand what Price Qty Multiplier is.

> In the same vain should we display the Floor and Cap in decimals representation and only behind the scenes modify them to integers.
> Perhaps displaying to the user the current price would be helpful as well

It makes sense. Without helper text, user will not know where is Floor and Cap come from. And it is error prone to have users convert price from ether to wei by themself.

eswarasai commented:
@dilatebrave -- PR looks good to me. I'll let @pelsasser decide if there are any other functional changes required here. Thanks :)

pelsasser commented:
@eswarasai good to merge by me.

I would recommend we create two new issues

1. remove the price qty multiplier so the user doesn't have to do anything with it 
2. display current price to user and then allow them to use decimal places for floor and cap, add some error checking to ensure they are reasonable (perhaps within 100% range of the current price) and then behind the scenes convert them to integers (or wei). 

@eswarasai - Can you create these?

Thanks for all the help here!

eswarasai commented:
@pelsasser -- Sure. Sounds good to me. I'll create a single issue as removing price qty multiplier can be dealt in the same one without much efforts.
